librosa/feature/spectral.py

```diff
@@ -849,19 +849,32 @@ def rms(y=None, S=None, frame_length=2048, hop_length=512,
                        frame_length=frame_length,
                        hop_length=hop_length)
 
-        # No normalization is necessary for time-domain input
-        norm = 1
+        # Calculate power
+        power = np.mean(np.abs(x)**2, axis=0, keepdims=True)
     elif S is not None:
-        x, _ = _spectrogram(y=y, S=S,
-                            n_fft=frame_length,
-                            hop_length=hop_length)
-
-        # FFT introduces a scaling of n_fft to energy calculations
-        norm = 2 * (x.shape[0] - 1)
+        # Check the frame length
+        if S.shape[0] != frame_length // 2 + 1:
+            raise ParameterError(
+                    'Since S.shape[0] is {}, '
+                    'frame_length is expected to be {} or {}; '
+                    'found {}'.format(
+                            S.shape[0],
+                            S.shape[0] * 2 - 2, S.shape[0] * 2 - 1,
+                            frame_length))
+
+        # power spectrogram
+        x = np.abs(S) ** 2
+
+        # Recover negative frequency bins' power
+        end = None if frame_length % 2 else -1
+        x[1:end] *= 2
```

Comment:
I think this could be done with fewer multiply adds if we scale `x[0] *= 0.5` (and `x[-1]` for even length) and then multiply `power = 2 * np.sum(...)`.

Reply:
I never came up with such a smart calculation. Thank you, I have reflected the proposed algorithm.